tests/xdc/history_replication_dlq_test.go

```diff
@@ -0,0 +1,454 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package xdc
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"os"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/gogo/protobuf/jsonpb"
+	"github.com/pborman/uuid"
+	"github.com/stretchr/testify/suite"
+	"github.com/urfave/cli/v2"
+	"go.temporal.io/api/serviceerror"
+	"go.temporal.io/api/workflowservice/v1"
+	sdkclient "go.temporal.io/sdk/client"
+	sdkworker "go.temporal.io/sdk/worker"
+	"go.temporal.io/sdk/workflow"
+	"go.uber.org/fx"
+
+	enumspb "go.temporal.io/server/api/enums/v1"
+	replicationspb "go.temporal.io/server/api/replication/v1"
+	"go.temporal.io/server/common/dynamicconfig"
+	"go.temporal.io/server/common/namespace"
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/primitives"
+	"go.temporal.io/server/common/primitives/timestamp"
+	"go.temporal.io/server/service/history/replication"
+	"go.temporal.io/server/service/history/tasks"
+	"go.temporal.io/server/tests"
+	"go.temporal.io/server/tools/tdbg"
+)
+
+// This file contains tests for the history replication DLQ feature. It uses a faulty replication task executor to force
+// replication tasks to be added to the DLQ, and then uses the tdbg CLI to read the replication tasks from the DLQ and
+// verify that they are correct.
+
+type (
+	historyReplicationDLQSuite struct {
+		xdcBaseSuite
+		replicationTaskExecutorParams          replicationTaskExecutorParams
+		executionManagerParams                 executionManagerParams
+		namespaceReplicationTaskExecutorParams namespaceReplicationTaskExecutorParams
```

Comment:
just to confirm we decorate the namespace task executor to make sure the namespace is replicated to the target cluster before starting replication?

Reply:
Yeah, that's exactly what it's fore. It's invasive, but I considered 2 alternatives, and both had drawbacks that I figured were worse:
1. **Sleep:** We can just sleep after registering the namespace. I don't like this because it slows down the test a lot, and it's still a race condition.
2. **Retry:** We can connect to the standby cluster and keep retrying `DescribeNamespace` until it doesn't return a not-found error. This means there's no race condition. However, it would be slower because of the sleeps between attempts and network overhead. It would also be inefficient because of the retries themselves. Finally, we would also need to come up with a good retry policy.